src/Microsoft.Health.Fhir.Shared.Api/Features/Filters/ValidationQueryFilterAndParameterParserAttribute.cs

```diff
@@ -0,0 +1,149 @@
+using System.Collections.Generic;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Api.Features.Routing;
+using Microsoft.Health.Fhir.Core.Features;
+using Microsoft.Health.Fhir.Core.Features.Operations;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Features.Validation;
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Api.Features.Filters
+{
+    [AttributeUsage(AttributeTargets.Class)]
+    internal class ValidationQueryFilterAndParameterParserAttribute : ActionFilterAttribute
+    {
+        private readonly FeatureConfiguration _features;
+
+        public ValidationQueryFilterAndParameterParserAttribute(IOptions<FeatureConfiguration> features)
+        {
+            EnsureArg.IsNotNull(features, nameof(features));
+            EnsureArg.IsNotNull(features.Value, nameof(features));
+
+            _features = features.Value;
+        }
+
+        public override void OnActionExecuting(ActionExecutingContext context)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+
+            if (!_features.SupportsValidate)
+            {
+                throw new OperationNotImplementedException(Api.Resources.ValidationNotSupported);
+            }
+
+            var idMode = context.ActionDescriptor.DisplayName.Contains("ValidateById", StringComparison.OrdinalIgnoreCase);
+
+            string mode = null;
+            StringValues modes;
+            if (context.HttpContext.Request.Query.TryGetValue(KnownQueryParameterNames.Mode, out modes) && modes.Count > 0)
+            {
+                if (modes.Count > 1)
+                {
+                    throw new BadRequestException(Api.Resources.MultipleModesProvided);
+                }
+
+                mode = modes[0];
+            }
+
+            string profile = null;
+            StringValues profiles;
+            if (context.HttpContext.Request.Query.TryGetValue(KnownQueryParameterNames.Profile, out profiles) && profiles.Count > 0)
+            {
+                if (profiles.Count > 1)
+                {
+                    throw new BadRequestException(Api.Resources.MultipleProfilesProvided);
+                }
+
+                profile = profiles[0];
+            }
+
+            if (context.RouteData.Values.TryGetValue(KnownActionParameterNames.ResourceType, out var actionModelType) &&
+               context.ActionArguments.TryGetValue(KnownActionParameterNames.Resource, out var parsedModel))
+            {
+                if (((Resource)parsedModel).ResourceType == ResourceType.Parameters)
+                {
+                    var resource = ParseParameters((Parameters)parsedModel, ref profile, ref mode);
+
+                    context.ActionArguments[KnownActionParameterNames.Resource] = resource;
```

Comment:
Just to confirm what's happening here, we want to replace the original resource in the request (which is a Paramter resource) with the resource that was included as part of the Parameter resource.

Reply:
I saw this as well, I think its so it validates with the `ValidateResourceTypeFilter`.

Perhaps a better option is to update the filter to allow this? `ValidateResourceTypeFilter(AllowParametersResource=true)`?

Reply:
That would allow the type filter to pass. I will also need to do this for the id filter, as it expects the id to be at the top level. My original idea with switching the value of the resource was so that as few areas as possible would need to know about whether it wrapped in a parameter resource.

Reply:
If there are a lot of exceptions you're working around, I'd remove the Filter from the Controller level. It could be applied at the Action level if it makes sense.

Also, we could split the actions so a regular resource and Parameter resource go to different action methods, this could be done with IActionSelector.